contracts/SpritzPayV1.sol

```diff
@@ -66,14 +84,228 @@ contract SpritzPayV1 is SpritzPayStorage, Initializable, OwnableUpgradeable, Pau
+            if (!refundSourceTokenSuccess) {
+                revert FailedRefund({ tokenAddress: sourceTokenAddress, amount: remainingBalance });
+            }
+        }
+
+        if (address(this).balance > startingBalance) {
+            //slither-disable-next-line arbitrary-send
+            payable(_msgSender()).transfer(startingBalance - address(this).balance);
+        }
+
+        console.log(
+            "after refunds: source-%s, payment-%s, native-%s",
+            isNativeSwap ? 0 : sourceToken.balanceOf(address(this)),
+            paymentToken.balanceOf(address(this)),
+            address(this).balance
+        );
+
+        logPayment(sourceTokenAddress, sourceTokenAmount, paymentTokenAddress, paymentTokenAmount, paymentReference);
+    }
+
+    /**
+     * @notice Pay by swapping token or with native currency, using 0x as dex aggregator
+     * @param sourceTokenAddress Address of the token being sold for payment
+     * @param sourceTokenAmount Amount of the token being sold for payment
+     * @param paymentTokenAddress Address of the target payment token
+     * @param paymentTokenAmount Amount of the target payment token
+     * @param paymentReference Reference of the payment related
+     */
+    function payWithUniswap(
+        address sourceTokenAddress,
+        uint256 sourceTokenAmount,
+        address paymentTokenAddress,
+        uint256 paymentTokenAmount,
+        bytes32 paymentReference
+    ) external payable whenNotPaused nonReentrant {
+        bool isNativeSwap = sourceTokenAddress == address(0);
+
+        IERC20 sourceToken = IERC20(sourceTokenAddress);
+        IERC20 paymentToken = IERC20(paymentTokenAddress);
+        IUniswapV2Router02 router = IUniswapV2Router02(swapTarget);
+
+        // If swap involves non-native token
+        if (!isNativeSwap) {
+            //Ensure our contract gives sufficient allowance to swap target
+            if (sourceToken.allowance(address(this), swapTarget) < sourceTokenAmount) {
+                bool approveSuccess = approveTokenSpend(sourceTokenAddress, swapTarget);
+                require(approveSuccess, "Could not approve swapTarget");
+            }
+
+            //Transfer from user to our contract
+            bool transferInSuccess = safeTransferToken(
+                sourceTokenAddress,
+                _msgSender(),
+                address(this),
+                sourceTokenAmount
+            );
+
+            if (!transferInSuccess) {
+                revert FailedTokenTransfer({
+                    tokenAddress: sourceTokenAddress,
+                    to: address(this),
+                    amount: sourceTokenAmount
+                });
+            }
+        }
+
+        address[] memory path = new address[](2);
+        path[0] = sourceTokenAddress;
+        path[1] = paymentTokenAddress;
+
+        router.swapTokensForExactTokens(paymentTokenAmount, sourceTokenAmount, path, paymentRecipient, block.timestamp);
```

Comment:
I think the return value shows how much it spent (somewhere in there). Can use that to determine how much to return